storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -109,89 +106,204 @@ impl RestoreCoordinator {
+        ensure!(
+            snapshot_before_lhs.is_some() && snapshot_before_target.is_some(),
+            "No snapshot exists before the target version({}) including genesis",
+            target_version
+        );
+        let snapshot_before_lhs = snapshot_before_lhs.unwrap();
+        let snapshot_before_target = snapshot_before_target.unwrap();
+        ensure!(
+            snapshot_before_lhs.version <= snapshot_before_target.version,
+            "snapshot_before_target {} should be larger than or equal to snapshot_before_lhs {}",
+            snapshot_before_target.version,
+            snapshot_before_lhs.version
+        );
+
+        // Two flags for resuming from a previous in-progress restore
+        // Expected version can be used to tell where to resume when applying writesets or replaying txns
+        // Tree restore in progress can be used to tell the status of 2nd snapshot restore
+        let expected_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
-            info!(
-                next_txn_version = next_txn_version,
-                "DB is ready to accept transactions, start the node to catch up with the chain. \
-                If the node is unable to catch up because the DB is too old, delete the data folder \
-                and bootstrap again.",
-            );
-            return Ok(());
+        let tree_restore_in_progress = self
+            .global_opt
+            .run_mode
+            .get_state_leaf_count(snapshot_before_target.version)
+            > 0;
+
+        info!(
+            lhs = lhs,
+            target_version = target_version,
+            "Starting restore DB from version {} to {}, snapshot_before_lhs: {}, snapshot_before_target: {}, tree restore in progress: {}, expected_version: {} \n\
+            Note: we only guarantee resume from previous in-progress restore. If you want to restore a new DB, please either specify a new target db dir or delete previous in-progress DB in the target db dir.
+            ",
+            lhs,
+            target_version,
+            snapshot_before_lhs.version,
+            snapshot_before_target.version,
+            tree_restore_in_progress,
+            expected_version,
+        );
+        let transaction_backups = metadata_view
+            .select_transaction_backups(snapshot_before_lhs.version, target_version)?;
+        let epoch_ending_backups = metadata_view.select_epoch_ending_backups(target_version)?;
+        let mut expected_txn_history_so_far = None;
+
+        // Restore the the state kv between lhs and rs
+        // Ensure the expected_version is smaller than target_version in case we want to resume from in-progress restore
+        if snapshot_before_lhs.version < snapshot_before_target.version
+            && expected_version <= snapshot_before_target.version
```

Comment:
I think as long as there's already ledger history in the db, we don't need to recover the kv-only snapshot, instead, we just  go on recovering the ledger.  The only bad case I can think of is that the db_next_version is very much earlier than start_version, we can just print it out.

Move the humongous log line (your execution plan) after you know exactly what you want to do: what's the first version you are actually gonna save, which exact snapshots you are gonna use, etc.